### PR TITLE
Support non-existing proofs in RwTable

### LINKED ISSUE #895: Support non-existing proofs in RwTable for circuit
Since @ed255 has already implemented it in Spec PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/291 👍. I suppose if I could try to move it to our Circuit code if no one is working on this (then update the `BALANCE` PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/683). Thanks.

### PR DESCRIPTION
**Pull request by silathdiir**
Close https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/895

Supposed to move https://github.com/privacy-scaling-explorations/zkevm-specs/pull/291 from spec to circuit.

### CODE REVIEW

**Review by miha-stopar — APPROVED** [review]
LGTM!

**Comment by ChihChengLiang** [comment]
@z2trillion Can you take a look into this PR when you have time?

File: zkevm-circuits/src/state_circuit.rs [hunk 1]
```diff
@@ -41,14 +41,21 @@ const N_LIMBS_ID: usize = 2;
 /// Config for StateCircuit
 #[derive(Clone)]
 pub struct StateCircuitConfig<F> {
-    selector: Column<Fixed>, // Figure out why you get errors when this is Selector.
```
[the remark this answers is not in the document]
Comment (z2trillion) [reply zkevm-circuits/src/state_circuit.rs:44]: I originally left this comment. Gates require a fixed column, but `Selector` is a fixed column. I think the check in halo2 may not have taken this into account, but I have not investigated deeper.

File: zkevm-circuits/src/state_circuit.rs [hunk 2]
```diff
@@ -41,14 +41,21 @@ const N_LIMBS_ID: usize = 2;
...
-                                    * and Rw::AccountStorage rows this is the committed value in
-                                    * the MPT, for others, it is 0. */
+    // Assigned value at the start of the block. For Rw::Account and
+    // Rw::AccountStorage rows this is the committed value in the MPT, for
+    // others, it is 0.
+    initial_value: Column<Advice>,
+    // For Rw::AccountStorage, identify if committed value or new value is zero.
+    // Will do lookup for ProofType::StorageDoesNotExist if both are zero,
+    // otherwise do lookup for ProofType::StorageChanged (either is non-zero).
+    is_initial_value_zero: Column<Advice>,
```
Comment (z2trillion) [inline zkevm-circuits/src/state_circuit.rs:57]: Can you combine these two with a `BatchedIsZeroGadget`?

**Review by z2trillion — CHANGES_REQUESTED** [review]

File: zkevm-circuits/src/state_circuit.rs (hunk 2 above)
Comment (z2trillion) [reply zkevm-circuits/src/state_circuit.rs]: It would be better to use the `BatchedIsZeroGadget` in zkevm-circuits/src/evm_circuit/util/math_gadget.rs.

Comment (silathdiir) [reply zkevm-circuits/src/state_circuit.rs]: I tried to replace this field with `BatchedIsZeroGadget`, but it seems that the first parameter of function [BatchedIsZeroGadget::construct](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/util/math_gadget.rs#L102) is `evm_circuit::constraint_builder::ConstraintBuilder` (not `state_circuit::constraint_builder::ConstraintBuilder`).

And it seems that the `state_circuit::constraint_builder::ConstraintBuilder` needs this `StateCircuitConfig` (this field belongs to) to [construct here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/state_circuit.rs#L135).

@z2trillion  Any suggestion to refactor `BatchedIsZeroGadget` or `state_circuit::constraint_builder::ConstraintBuilder`? Thanks.

Comment (z2trillion) [reply zkevm-circuits/src/state_circuit.rs]: Sorry about this. I didn't realize that `BatchedIsZeroGadget` requires the evm `ConstraintBuilder` to construct. Can you just add a todo here: "TODO: use BatchedIsZeroGadget here once it no longer requires the evm circuit constraint builder."

**Review by z2trillion — APPROVED** [review]

File: zkevm-circuits/src/state_circuit.rs (hunk 2 above)
Comment (silathdiir) [reply zkevm-circuits/src/state_circuit.rs]: Yes. I see. Thanks 🙏.
I will add `TODO: use BatchedIsZeroGadget here once it no longer requires the evm circuit constraint builder.` in the `BALANCE` circuit PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/683 (which will use this non-existing proof code).

Comment (silathdiir) [reply zkevm-circuits/src/state_circuit.rs]: I saw the PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/961. Thanks.

**Pull request merged**